### ISSUE TITLE
canvas.js, RadioButton, random()?

### ISSUE THREAD

**Issue by DerozerDSP** [ryukau/UhhyouWebSynthesizers#6]
Hi dude,

I'm trying to add random() function on your RadioButton within canvas.js file (i.e. here for example https://github.com/ryukau/UhhyouWebSynthesizers/blob/main/oldsynth/SingenBD2/canvas.js#L240).

Tried this way, but seems it doesnt update value and gui:

`
random() {
	var randomIndex = Math.floor(Math.random() * this.buttons.length);
	const randomButton = this.buttons[randomIndex];
	randomButton.setAttribute('checked', true);
	this.value = randomButton.value;
}

What's the correct way of doing this?
Thanks for any tips and/or suggestions!!
`

**Comment by ryukau** [comment]
4 spam comments above were deleted.

@DerozerDSP I'll write a proper reply a bit later.

**Comment by DerozerDSP** [comment]
@ryukau I see, don't worry :) See you later, thanks.

**Comment by ryukau** [comment]
I added a randomization of FM type in 76e4d9e.

It seems like radio buttons can't be visually selected from JavaScript. The workaround I used was to delete the all existing radio buttons and adding them again to the container, which is `RadioButton.divButtons`.

@DerozerDSP Are you using SingenBD2 frequently? If you'd like to extend it further more, I'll consider porting it to the new style of code at some point, but not anytime soon.

**Comment by ryukau** [comment]
Okay, then I leave SingenBD2 as is. And thank you for the compliment.

On programming in JavaScript, maybe a key point is that JavaScript can make any binary data through [typed arrays](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Typed_arrays). It makes possible to generate WAVE or any other file formats.

**Comment by ryukau** [comment]
To me, it seems like working as intended.

When "Random" is pressed, It's calling `RadioButton.refresh()` in canvas.js, then calling `refresh()` in index.js. I confirmed this by putting `console.log()` on following places:

- [refresh() in index.js](https://github.com/ryukau/UhhyouWebSynthesizers/blob/76e4d9ed1f50fbb4f54077853d4f11c27f06d39f/oldsynth/SingenBD2/index.js#L237)
- [refresh() in canvas.js](https://github.com/ryukau/UhhyouWebSynthesizers/blob/76e4d9ed1f50fbb4f54077853d4f11c27f06d39f/oldsynth/SingenBD2/canvas.js#L252)

I used Chrome and Firefox on Windows.

Just in case, force refreshing the page might solve the issue if you haven't.

If the problem is not what I'm talking, I need more detailed steps to reproduce the bug. Or, a video that shows the bug. (I use [OBS](https://obsproject.com/) for screen capture).

**Comment by DerozerDSP** [comment]
I meant: it doesn't recall the refresh() function linked to radioButtonModulationType on index.js:

<img width="562" alt="image" src="https://github.com/user-attachments/assets/9b106104-d892-402e-b4a9-baf2dc3bf568">

Try to random, look the Waveform, and than manually (if for example is selected Pow FM) to click on Muffled FM and than again Pow FM: the results look wave are different from the initial Pow FM...

**Comment by ryukau** [comment]
Ah, I get it. That's by design.

When "Random" button is pressed, it calls `random()` in index.js. Then `random()` calls `refresh()` only once.

`random()` doesn't call the event callbacks set on the GUI widgets. The function you are pointing to is one of those callbacks. Otherwise, `refresh()` will be repeatedly called as much as the number of widgets on the page, which is inefficient.

**Comment by ryukau** [comment]
Waveform looks okay on my environment.

https://github.com/user-attachments/assets/7c59802c-9624-49fb-ad8e-0e7e7bf5307d

**Comment by DerozerDSP** [comment]
I see what's the problem. Basically I've wrap the function within "random" also for RadioButton, such as:

`
  add(label, checked = undefined) {
    var input = document.createElement("input")
    input.type = "radio"
    input.name = this.name
    input.value = label
    if ((this.buttons.length <= 0 && checked === undefined) || checked === true) {
      input.setAttribute("checked", true)
      this.value = label
    }
    this.buttons.push(input)

    var labelElement = document.createElement("label")
    labelElement.addEventListener("change", (event) => this.onChange(event), false)
    labelElement.innerHTML = input.outerHTML + label
    this.divButtons.appendChild(labelElement)
  }

  random() {
    var randomIndex = Math.floor(Math.random() * this.buttons.length);
    let values = this.buttons.map(radio => radio.value);

    this.divButtons.innerHTML = "";
    this.buttons = [];
    for (let idx = 0; idx < values.length; ++idx) {
      this.add(values[idx], idx === randomIndex);
    }
  }
`

but this way, outside (on index.js), fmType won't be refreshed.

Probably the faster way is to do `return randomIndex;` on random() function, and when calling it, assign the returned value, such as `fmType = radioButtonModulationType.random();`

Not so elegant, but it works for now :)

**Comment by ryukau** [comment]
Your code is much better, because it's not leaking the implementation detail of RadioButton to the outside.

Some notes:

- `fmType` global variable on index.js is a bad design by me.
- Radio button (`<input type="radio">`) is not suitable for this use case. It's better to use `<select>` as in `common/gui/combobox.js`.

These kind of things are the reason they are called oldsynth.

I'll close this issue as it seems to be solved. If there's anything, feel free to add comment or open another issue.
